### DISCUSSION TITLE
Migrating from the old operator to the new one

### DISCUSSION THREAD

**Discussion by philomory** [keycloak/keycloak#12139] [Archive]
Once the new operator is fully GA in Keycloak 20, do we have any idea what the migration path will look like from the legacy operator? I'm particularly interested migrating installations which make heavy use of the client CRD, including the automatic secret injection they provide. I'm worried that the transition from the legacy operator to the new operator may be onerous enough that we're better off just not using Keycloak at all until 20 releases and the new operator is GA.

**Comment by andreaTP** [comment]
Hi @philomory ! Thanks for your interest in the new operator!

> including the automatic secret injection they provide.

the new operator does support secret injection in most of the relevant places, would mind checking if anything is missing from your perspective?

> client CRD

Currently, the new operator supports only full Realm bulk import, I'm confident that you can migrate the client CRs to a realm CR (containing the clients). Please, try it out, and report if you face any technical limitations.

**Reply by philomory** [reply]
Hi @andreaTP 
> the new operator does support secret injection in most of the relevant places, would mind checking if anything is missing from your perspective?

How would this actually work? There's no documentation about it. I'm referring to the feature in the legacy operator where, when you create a KeycloakClient CR, in addition to creating the client resource within Keycloak itself, it also creates a Kubernetes Secret object with the credentials for that client, consumable from various Kubernetes workloads. Per [the documentation for the legacy operator](https://www.keycloak.org/docs/latest/server_installation/index.html#_client-cr):
> After a client is created, the Operator creates a Secret containing the Client ID and the client’s secret using the following naming pattern: `keycloak-client-secret-<custom resource name>`

Does this functionality exist in the new operator? If so, I don't see that documented anywhere. With no KeycloakClient CR, how does it know what namespace to put the Secret object in? Does it just put it in the same namespace as the Keycloak installation itself?

>Currently, the new operator supports only full Realm bulk import, I'm confident that you can migrate the client CRs to a realm CR (containing the clients). Please, try it out, and report if you face any technical limitations.

As stated by @jbpin in [another thread](https://github.com/keycloak/keycloak/discussions/11724), this isn't really a usable solution. I'm not looking to _import_ a realm as a one-time operation, I'm looking to _manage_ the realm using Kubernetes CRs (something you can do with the legacy operator). My question at this point is, if I start by using the legacy operator, will I eventually be able to migrate that configuration into the new operator _and keep managing it the same way_, or am I better off just not using Keycloak Operator _at all_ until 20.0 is GA, to avoid dealing with a messy, painful migration down the line?
